File: Lecture 4/FinalCode - Java/app/src/main/java/com/inovaeg/androidtrainingpharos2020/dataStorage/BaseStorageActivity.java
package com.inovaeg.androidtrainingpharos2020.dataStorage;

import android.os.Bundle;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.inovaeg.androidtrainingpharos2020.R;

import java.util.ArrayList;
import java.util.List;

public class BaseStorageActivity extends AppCompatActivity {
    public static final String TYPE_SHARED_PREF = "TYPE_SHARED_PREF";
    public static final String TYPE_DATABASE = "TYPE_DATABASE";
    public static final String TYPE_FILE_SAVE = "TYPE_FILE_SAVE";
    public static final String TYPE_KEY = "TYPE_KEY";
    private List<String> names = new ArrayList<>();
    private EditText editText;
    private TextView tvNames;
    private String type;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getIntent().getExtras() != null && getIntent().getExtras().getString(TYPE_KEY) != null) {
            type = getIntent().getExtras().getString(TYPE_KEY);
        }
        setContentView(R.layout.activity_shared_pref);
        editText = findViewById(R.id.ed_name);
        tvNames = findViewById(R.id.tv_names);
        findViewById(R.id.bt_add).setOnClickListener(v -> {
            if (editText.getText() != null && editText.getText().length() > 0) {
                names.add(editText.getText().toString());
            }
            tvNames.setText(joinInSingleString(names));
            editText.setText("");
        });

    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        switch (type) {
            case TYPE_SHARED_PREF: {
                saveNamesInSharedPref(names);
                break;
            }
            case TYPE_DATABASE: {
                saveNamesInDatabase(names);
                break;
            }
            case TYPE_FILE_SAVE: {
                saveNamesInFile(names);
                break;
            }
        }
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        switch (type) {
            case TYPE_SHARED_PREF: {
                setNames(restoreFromSharedPref());
                break;
            }
            case TYPE_DATABASE: {
                setNames(restoreFromDatabase());
                break;
            }
            case TYPE_FILE_SAVE: {
                setNames(restoreFromFile());
                break;
            }
        }
        findViewById(R.id.bt_add).callOnClick();
    }

    private void setNames(List<String> names) {
        if (names != null) {
            this.names = names;
        }
    }

    //    sharedPref methods
    protected void saveNamesInSharedPref(List<String> names) {

    }

    protected List<String> restoreFromSharedPref() {

        return null;
    }

    ///////////////
    //////////file
    protected void saveNamesInFile(List<String> names) {

    }

    protected List<String> restoreFromFile() {

        return null;
    }
    //////////////

    //DataBase
    protected void saveNamesInDatabase(List<String> names) {
    }

    protected List<String> restoreFromDatabase() {
        return null;
    }
    ////////////

    final protected String joinInSingleString(List<String> names) {
        StringBuilder joinedNames = new StringBuilder();
        for (String name : names
        ) {
            joinedNames.append(name);
            joinedNames.append(" ");
        }
        return joinedNames.toString();
    }
}

File: Lecture 4/FinalCode - Kotlin/app/src/main/java/com/inovaeg/androidtrainingpharos2020/dataStorage/BaseStorageActivity.kt
package com.inovaeg.androidtrainingpharos2020.dataStorage

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.inovaeg.androidtrainingpharos2020.R
import kotlinx.android.synthetic.main.activity_shared_pref.*
import java.util.*

open class BaseStorageActivity : AppCompatActivity() {
    private var names: MutableList<String> = ArrayList()
    private var type: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (intent.extras != null && intent.extras!!.getString(TYPE_KEY) != null) {
            type = intent.extras!!.getString(TYPE_KEY)
        }
        setContentView(R.layout.activity_shared_pref)

        findViewById<View>(R.id.bt_add).setOnClickListener { v: View? ->
            if (ed_name.getText() != null && ed_name.getText().length > 0) {
                names.add(ed_name.getText().toString())
            }
            tv_names.setText(joinInSingleString(names))
            ed_name.setText("")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        when (type) {
            TYPE_SHARED_PREF -> {
                saveNamesInSharedPref(names)
            }
            TYPE_DATABASE -> {
                saveNamesInDatabase(names)
            }
            TYPE_FILE_SAVE -> {
                saveNamesInFile(names)
            }
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        when (type) {
            TYPE_SHARED_PREF -> {
                setNames(restoreFromSharedPref())
            }
            TYPE_DATABASE -> {
                setNames(restoreFromDatabase())
            }
            TYPE_FILE_SAVE -> {
                setNames(restoreFromFile())
            }
        }
        findViewById<View>(R.id.bt_add).callOnClick()
    }

    private fun setNames(names: MutableList<String>?) {
        if (names != null) {
            this.names = names
        }
    }

    //    sharedPref methods
    protected open fun saveNamesInSharedPref(names: List<String>?) {}
    protected open fun restoreFromSharedPref(): MutableList<String>? {
        return null
    }

    ///////////////
    //////////file
    protected open fun saveNamesInFile(names: List<String>?) {}
    protected open fun restoreFromFile(): MutableList<String>? {
        return null
    }

    //////////////
    //DataBase
    protected open fun saveNamesInDatabase(names: List<String>) {}
    protected open fun restoreFromDatabase(): MutableList<String>? {
        return null
    }

    ////////////
    protected fun joinInSingleString(names: List<String?>): String {
        val joinedNames = StringBuilder()
        for (name in names) {
            joinedNames.append(name)
            joinedNames.append(" ")
        }
        return joinedNames.toString()
    }

    companion object {
        const val TYPE_SHARED_PREF = "TYPE_SHARED_PREF"
        const val TYPE_DATABASE = "TYPE_DATABASE"
        const val TYPE_FILE_SAVE = "TYPE_FILE_SAVE"
        const val TYPE_KEY = "TYPE_KEY"
    }
}